docs/proposals/advanced-nginx-extensions.md

```diff
@@ -26,3 +74,915 @@ NGINX use cases. To allow them to implement those use cases, we need to bring a
+values.
+
+Next, Cluster operator defines the following CRD for those values. Note that the values are validated to prevent invalid
+or
+malicious values from being propagated into the NGINX config:
+
+```go
+// RateLimitingPolicy is a Direct Attached Policy to configure rate-limiting of client requests.
+type RateLimitingPolicy struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of the ClientSettingsPolicy.
+	Spec RateLimitingPolicySpec `json:"spec"`
+
+	// Status defines the state of the ClientSettingsPolicy.
+	Status gatewayv1alpha2.PolicyStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// RateLimitingPolicyList contains a list of RateLimitingPolicies.
+type RateLimitingPolicyList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []ClientSettingsPolicy `json:"items"`
+}
+
+// RateLimitingPolicySpec defines the desired state of RateLimitingPolicy.
+type RateLimitingPolicySpec struct {
+	// The rate of requests permitted per second.
+	//
+	// +kubebuilder:validation:Minimum=1
+	Rate int32 `json:"rate"`
+
+	// Burst delays excessive requests until their number exceeds the burst value, in which case the request is
+	// terminated with an error.
+	//
+	// +optional
+	// +kubebuilder:validation:Minimum=1
+	Burst *int32 `json:"burst,omitempty"`
+
+	// TargetRefs identifies the API object(s) to apply the policy to.
+	// Objects must be in the same namespace as the policy.
+	// Support: HTTPRoute.
+	//
+	// +kubebuilder:validation:MaxItems=16
+	// +kubebuilder:validation:XValidation:message="TargetRef Kind must be: HTTPRoute.",rule="self.exists(t, t.kind=='HTTPRoute')"
+	// +kubebuilder:validation:XValidation:message="TargetRef Group must be gateway.networking.k8s.io.",rule="self.all(t, t.group=='gateway.networking.k8s.io')"
+	//nolint:lll
+	TargetRefs []gatewayv1alpha2.LocalPolicyTargetReference `json:"targetRefs"`
+}
+```
+
+Next, Cluster operator generates a CRD manifest from that go code and creates the CRD in the cluster.
+
+Next, Cluster operator creates templates that generate the NGINX configuration based on the fields of the CRD:
+
+- For http context:
+
+  ```text
+  {{ $rate := index $.Data "rate" }}
+  {{ $zoneName := $.Metadata.UID }}
+  limit_req_zone $binary_remote_addr zone={{ $zoneName }}:10m rate={{ $rate }}r/s;
+  ```
+
+- For location context:
+
+  ```text
+  {{ $zoneName := $.Metadata.UID }}
+  {{ $burst := index $.Data "burst }}
+  limit_req zone={{ $zoneName }}{{ if $burst }}burst={{ $burst }}{{ end }};
+  ```
+
+Next, Cluster operator allows Application developer to use the feature by creating a SnippetsTemplate, which binds
+the CRD with the templates:
+
+```yaml
+apiVersion: gateway.nginx.org/v1alpha1
+kind: SnippetsTemplate
+metadata:
+  name: expose
+spec:
+  valuesCRD: ratelimitingpolicy
```

Comment:
Does this need to be a GVK?

Reply:
reading the CRD will get that info. For example, https://github.com/nginxinc/nginx-gateway-fabric/blob/main/config/crd/bases/gateway.nginx.org_clientsettingspolicies.yaml

Potentially, we could read fields of that CRD - for example, check if it has the right structure (targetFields, spec and status). but this might be an overkill 

at the same time, NGF will stick need check for those fields - when parsing watched CRs of that CRD and updating their statuses

Reply:
Don't we need the group and version to GET the CRD from the k8s api?

Reply:
> Don't we need the group and version to GET the CRD from the k8s api?

just the name:
```
kubectl get crd nginxproxies.gateway.nginx.org -o yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  annotations:
    controller-gen.kubebuilder.io/version: v0.15.0
  . . .
```

Reply:
That name includes the group though.

Reply:
What if there are two CRDs: 

`ratelimitingpolicy.team1.org`
`ratelimitingpolicy.team2.org`

how we will know which one to look up?

Reply:
> That name includes the group though.

yep

https://github.com/kubernetes/kubernetes/blob/b977096f5a23916bf0ff0af4a6ee6932b37606a9/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L78-L81


> What if there are two CRDs:
> ratelimitingpolicy.team1.org
> ratelimitingpolicy.team2.org
> how we will know which one to look up?

I might be missing something.
The CRD name uniquely identify the CRD. The group and version can be looked up form it. For example,  https://github.com/nginxinc/nginx-gateway-fabric/blob/0edd3a819b249827f4c6ae4f806232d1c220d162/config/crd/bases/gateway.nginx.org_clientsettingspolicies.yaml#L11-L22

At the same time, there could be multiple versions of CRD. For example:
- https://github.com/kubernetes-sigs/gateway-api/blob/798801bd0af80f2585e26e82a483f5648568b6cd/config/crd/standard/gateway.networking.k8s.io_httproutes.yaml#L28
- https://github.com/kubernetes-sigs/gateway-api/blob/798801bd0af80f2585e26e82a483f5648568b6cd/config/crd/standard/gateway.networking.k8s.io_httproutes.yaml#L3044

so if NGF just knows CRD name, it will have to chose which one to use. It can choose to use the one that is served.

However, with your suggestion, we can explicitly configure version, which seems more flexible.

Reply:
Ok, so the `valuesCRD` must contain the full CRD name, not just `ratelimiitingpolicy` like in the example. 

Can we make that clear in the proposal?

> so if NGF just knows CRD name, it will have to chose which one to use. It can choose to use the one that is served.

This should be fine 👍